### PR TITLE
memoize the current_user? method.

### LINKED ISSUE #32: Multiple calls to current_user per action
This is a minor thing, but something to think about. Here's a log showing which method is being called on a basic install with Authentic:

```text
web    | GET /
web    |  ▸ Handled by Home::Index
web    |  ▸ Ran protect_from_forgery
web    |  ▸ Ran test_backdoor
web    |  ▸ current_user?
web    |  ▸ current_user?
web    |  ▸ Rendered Lucky::WelcomePage
web    |  ▸ Sent 200 (1.56ms)
web    |
web    | GET /sign_up
web    |  ▸ Handled by SignUps::New
web    |  ▸ Ran protect_from_forgery
web    |  ▸ Ran test_backdoor
web    |  ▸ current_user?
web    |  ▸ Ran redirect_signed_in_users
web    |  ▸ current_user?
web    |  ▸ Rendered SignUps::NewPage
web    |  ▸ Sent 200 (939.0µs)
web    |
web    | POST /sign_ups
web    |  ▸ Handled by SignUps::Create
web    |  ▸ Ran protect_from_forgery
web    |  ▸ Ran test_backdoor
web    |  ▸ current_user?
web    |  ▸ Ran redirect_signed_in_users
web    |  ▸ Sent 302 (75.63ms)
web    |
web    | GET /me
web    |  ▸ Handled by Me::Show
web    |  ▸ Ran protect_from_forgery
web    |  ▸ Ran test_backdoor
web    |  ▸ current_user?
web    |  ▸ find_current_user
web    |  ▸ Ran require_sign_in
web    |  ▸ current_user?
web    |  ▸ find_current_user
web    |  ▸ Rendered Me::ShowPage
web    |  ▸ Sent 200 (2.08ms)
```

Notice once we get to the `/me` route, it's calling the `find_current_user` methods twice. We can't memoize that method because it takes an argument, but if we can find a way to memoize something in here, that could reduce the load time just a tad more.

### PR DESCRIPTION
**Pull request by jwoertink**
Fixes #32

Since we include Lucky as a dependency of this library, we know that `memoize` exists at the `Object` level

https://github.com/luckyframework/lucky/blob/0fcd828851266b9bb5d2ca4695d0ebdca352a675/src/charms/object.cr#L3

### CODE REVIEW

**Comment by jwoertink** [comment]
oops. I forgot that I already opened an issue on this https://github.com/luckyframework/lucky/issues/1396

File: src/authentic/action_helpers.cr [hunk 1]
```diff
@@ -28,7 +28,7 @@ module Authentic::ActionHelpers(T)
   #
   # This method should *not* be overridden. If you want to require a current user,
   # override the `current_user` method (note no `?`).
-  def current_user? : T?
+  memoize def current_user? : T?
```
Comment (matthewmcgarvey) [inline src/authentic/action_helpers.cr]: The question mark means that it's not memoizable it seems?

**Review by matthewmcgarvey — CHANGES_REQUESTED** [review]

File: src/authentic/action_helpers.cr (hunk 1 above)
Comment (jwoertink) [reply src/authentic/action_helpers.cr]: Yeah, we can't memoize predicate methods. That would have to be fixed first before we can do that here... I guess I could implement some cheap memoizing on this method with an instance variable?

File: src/authentic/action_helpers.cr [hunk 2]
```diff
@@ -24,15 +24,20 @@ module Authentic::ActionHelpers(T)
...
-    if id
-      find_current_user(id)
-    end
+  def current_user? : T?
+    @__current_user ||= ->{
+      id = session.get?(SIGN_IN_KEY)
+      if id
+        find_current_user(id)
+      end
+    }.call
```
Comment (matthewmcgarvey) [inline src/authentic/action_helpers.cr]:
```suggestion
    @__current_user ||= begin
      id = session.get?(SIGN_IN_KEY)
      if id
        find_current_user(id)
      end
    end
```

This should do the job unless I'm missing something that needs the proc

**Review by matthewmcgarvey — CHANGES_REQUESTED** [review]

File: src/authentic/action_helpers.cr (hunk 2 above)
Comment (jwoertink) [reply src/authentic/action_helpers.cr]: That's fine too. My go to is always using a proc so I just default to that, but this works.

**Review by matthewmcgarvey — APPROVED** [review]

**Pull request merged**